File: Java/Data Structures & Algorithms Module - Java/Week8/Stack.java
/*
Stack.java
Billy Cussen 
28/10/2020
*/
import java.util.*; 

public class Stack<T> implements StackInterface<T>{

    Node<T> first;

    public static void main(String [] args){

        StackInterface<Integer> myStack = new Stack<Integer>();
        System.out.println("Is Empty: "+myStack.isEmpty());
        
        try{
            System.out.println("Size: "+myStack.size());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }

        try{
            System.out.println("Peek: "+myStack.peek());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }

        try{
            System.out.println("Pop: "+myStack.pop());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }

        myStack.push(1);
        myStack.push(2);
        myStack.push(3);
        myStack.push(4);
        myStack.push(5);
        myStack.push(6);
        myStack.push(7);

        System.out.println("Is Empty: "+myStack.isEmpty());
        System.out.println("Size: "+myStack.size());
        System.out.println("Peek: "+myStack.peek());
        System.out.println("Pop: "+myStack.pop());
        System.out.println("Size: "+myStack.size());

        System.out.println(myStack.toString());
        
        int size = myStack.size();
        for(int i = 0; i < size;i++){
            System.out.println("Pop: "+myStack.pop());
        }
        System.out.println("Is Empty: "+myStack.isEmpty());
        
        try{
            System.out.println("Size: "+myStack.size());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }

        try{
            System.out.println("Peek: "+myStack.peek());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }

        try{
            System.out.println("Pop: "+myStack.pop());
        }catch(NullPointerException e){
            System.out.println(e.getMessage());
        }
    }

    @Override
    public void push(T element){
        if(isEmpty()){
            first = new Node<T>(element, null);
        } else{
            first = new Node<T>(element, first);
        }
    }
    
    @Override
    public T pop(){
        if(isEmpty()){
            throw new NullPointerException("EXCEPTION: The Stack is empty, can't pop element!");
        } else{
            T element = first.element;
            first = first.next;
            return element;
        }
    }
    
    @Override
    public T peek(){
        if(isEmpty()){
            throw new NullPointerException("EXCEPTION: The Stack is Empty, can't peek element!");
        }
        return first.element;
    }
    
    @Override
    public int size(){
        if(isEmpty()){
            throw new NullPointerException("EXCEPTION: The Stack is Empty, can't get Size!");
        }
        return sizeHelper(first);
    }

    private int sizeHelper(Node<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.next);
    }
    
    @Override
    public boolean isEmpty(){
        return first == null;
    }

    @Override
    public String toString(){
        if(isEmpty()){
            throw new NullPointerException("EXCEPTION: The Stack is empty, cannot print elements!");
        }
        String result = "\n***STACK TRAVERSAL***\n";
        int position = 0;
        return toStringHelper(result, position, first);
    }

    private String toStringHelper(String result, int position, Node<T> current){
        if(current==null){
            result+="***END***\n";
            return result;
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringHelper(result,++position,current.next);
    }
}

interface StackInterface<T>{
    void push(T element);
    T pop();
    T peek();
    int size();
    boolean isEmpty();
}

class Node<T>{

    T element;
    Node<T> next;

    Node(T element, Node<T> next){
        this.element = element;
        this.next = next;
    }

}

File: Python/Data Structures & Algorithms Module - Python/Revision/Stacks/Stack.py
"""
Stack.py
Billy Cussen
07/01/2021
"""
from Node import Node

class Stack:

    def __init__(self):
        self.first = None

    def isEmpty(self):
        return self.first == None

    def size(self):
        if self.isEmpty():
            return 0
        return self.sizeHelper(self.first)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.next)

    def toString(self):
        if self.isEmpty():
            raise Exception("Cannot print Stack, it is empty!")
        else:
            result = "***STACK TRAVERSAL***\n"
            position = 1
            return self.toStringHelper(self.first, result, position)

    def toStringHelper(self, current, result, position):
        if current == None:
            result += "***END***"
            return result
        result+= "Position "+str(position)+", Element: "+str(current.element)+"\n"
        position = position + 1
        return self.toStringHelper(current.next, result, position)

    def peek(self):
        if self.isEmpty():
            raise Exception("Unable to peek, Stack is Empty!")
        return self.first.element

    def push(self, element):
        if self.isEmpty():
            self.first = Node(element, None)
        else:
            self.first = Node(element, self.first)

    def pop(self):
        if self.isEmpty():
            raise Exception("Unable to Pop, Stack is Empty!")
        else:
            element = self.first.element
            self.first = self.first.next
            return element